lib/types.js

```diff
@@ -197,7 +199,20 @@ class Type {
         return Type.forSchema(obj, opts);
       });
       if (!UnionType) {
-        UnionType = isAmbiguous(types) ? WrappedUnionType : UnwrappedUnionType;
+        // either automatic detection or we have a projection function
+        if (typeof opts.wrapUnions === 'function') {
+          // we have a projection function
+          try {
+          UnionType = typeof opts.wrapUnions(types) !== 'undefined'
```

Comment:
we could potentially store the generated function here already, then there would only be one call to `wrapUnions` for each set of types instead of two, if we assume generating the projection function could possibly be expensive.

Reply:
Let's refactor the code to only call the function once per union. It's more efficient and less surprising for callers.

For example we can add a third argument to `UnwrappedUnionType` to forward it (or `false` if it did not return a function). When the argument is `undefined`, `UnwrappedUnionType` should still call `opts.wrapUnion` to support direct calls.